### DISCUSSION TITLE
GitHub Classroom cannot find my template repository (repo is private in a different org, private forking is allowed)

### DISCUSSION THREAD

**Discussion by firasm** [community/community#136760] [GitHub Education]
### Hi everyone,

Title says it all - trying to figure out how to set the starter code, which is in a different org, with a private repo:

<img width="1545" alt="Screenshot 2024-08-25 at 2 45 50 AM" src="https://github.com/user-attachments/assets/184723c8-5019-456b-b973-306abd036ad4">

This works if the repo with the starter code is public, but not if it's private.

Any ideas? Am I missing a setting somewhere? I tried cancelling and restarting, to no avail.

<img width="934" alt="Screenshot 2024-08-25 at 2 47 47 AM" src="https://github.com/user-attachments/assets/7de105d9-cb7b-4166-9b0f-7ad2b93a3a98">

<img width="845" alt="Screenshot 2024-08-25 at 2 47 56 AM" src="https://github.com/user-attachments/assets/1070fbcb-1f8e-4fbc-b427-e9415301ab17">

**Comment by 4po** [comment]
Hello, @firasm ! 😄 

To ensure GitHub Classroom can find your template repository, follow these steps:

1. Ensure that the template repository is accessible by the GitHub Classroom application.
2. Add GitHub Classroom as a collaborator to your repository.
3. Configure the necessary permissions for organization-level access.
Here’s a high-level plan and corresponding code snippets to achieve these steps:

**Step 1**: Ensure Your Template Repository is Accessible
First, confirm if your repo is set to allow private forking and is in the correct organization.

**Step 2:** Add GitHub Classroom as a Collaborator
GitHub Classroom might need direct access to your repository to use it as a template. Add `github-classroom` as a collaborator to your private repository:

- Actions to do manually:
1. Navigate to your private repository on GitHub.
2. Click on `Settings`.
3. Select `Manage Access` from the sidebar.
4. Click `Invite a collaborator`.
5. Add `github-classroom` as a collaborator.

**Step 3**: Configure Organization-Level Access
Ensure GitHub Classroom has the necessary access at the organization level:

- Actions to do manually:
1. Go to your GitHub organization settings.
2. Select `Third-party access`.
3. Ensure GitHub Classroom is listed and has the necessary permissions.

Additional Code to Automate Adding GitHub Classroom as a Collaborator
If you need to automate the process of adding GitHub Classroom as a collaborator using GitHub's REST API, you can use the following JavaScript code with Node.js:

```javascript
const { Octokit } = require("@octokit/rest");
require("dotenv").config();

const octokit = new Octokit({
  auth: process.env.GITHUB_TOKEN,
});

async function addCollaborator(owner, repo, username, permission) {
  try {
    await octokit.repos.addCollaborator({
      owner: owner,
      repo: repo,
      username: username,
      permission: permission, // Can be 'pull', 'triage', 'push', or 'admin'
    });
    console.log(`Successfully added ${username} as a collaborator to the ${repo} repository.`);
  } catch (error) {
    console.error("Error adding collaborator:", error);
  }
}

// Replace these with your actual details
const owner = "your-organization"; // GitHub organization/user
const repo = "your-repo"; // Template repository name
const username = "github-classroom"; // GitHub Classroom username
const permission = "push"; // Necessary permission level

addCollaborator(owner, repo, username, permission);
```

- Instructions for the above script:
1. Ensure you have a `.env` file with the `GITHUB_TOKEN` defined.
2. Install the Octokit library using `npm install @octokit/rest`. 

Here's how to set up your `.env` file if you want :

``` 
GITHUB_TOKEN=your_github_personal_access_token
``` 

Push Errors and Branch Protection
Ensure you have the necessary access permissions to the repository, and if you encounter branch protection issues:

- Check the branch protection rules and make sure your changes comply with them.
- Ensure GitHub Classroom is added as an exception to these rules if necessary.

Refer to [GitHub's branch protection feature](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/defining-the-mergeability-of-pull-requests/about-protected-branches) for more information.

By following these steps, you should be able to make your private repository accessible to GitHub Classroom and resolve the issue. 👍🏻

**Reply by firasm** [reply]
Hello!

Thanks for the detailed response!

A few things:

- there doesn't seem to be a GitHub account with the username `github-classroom`. Is it perhaps GitHub-Classroom-Bot?![image](https://github.com/user-attachments/assets/e0627c96-05e2-402d-992d-a2ab3808cf74)

- I added the bot, but it says the invitation is pending (to the whole org) and needs to be accepted. Is this a manual process for the GH classroom process team?

- I wondered about how to add the GH Classroom application (without a script) because it doesn't seem to appear on the App Marketplace. I even tried creating a dummy classroom and linking to that org and it still doesn't appear in the list of third party apps. 

Thanks again. 

Firas
